core/src/main/scala-2.13+/cats/data/NonEmptyLazyList.scala

```diff
@@ -0,0 +1,455 @@
+
+  /**
+    * Left-associative fold using f.
+    */
+  final def foldLeft[B](b: B)(f: (B, A) => B): B =
+    toLazyList.foldLeft(b)(f)
+
+  /**
+    * Right-associative fold using f.
+    */
+  final def foldRight[B](z: B)(f: (A, B) => B): B =
+    toLazyList.foldRight(z)(f)
+
+  /**
+    * Left-associative reduce using f.
+    */
+  final def reduceLeft(f: (A, A) => A): A =
+    toLazyList.reduceLeft(f)
+
+  /**
+    * Apply `f` to the "initial element" of this LazyList and lazily combine it
+    * with every other value using the given function `g`.
+    */ // TODO not implemented in LazyList
+  final def reduceLeftTo[B](f: A => B)(g: (B, A) => B): B =
+    toLazyList.reduceLeftTo(f)(g)
+
+
+  /**
+    * Right-associative reduce using f.
+    */
+  final def reduceRight[AA >: A](f: (A, AA) => AA): AA =
+    toLazyList.reduceRight(f)
+
+  /**
+    * Apply `f` to the "initial element" of this NonEmptyLazyList and
+    * lazily combine it with every other value using the given function `g`.
+    */ // TODO not implemented in LazyList
+  final def reduceRightTo[B](f: A => B)(g: (A, B) => B): B =
+    toLazyList.reduceRightTo(f)(g)
+
+  /**
+    * Reduce using the Semigroup of A
+    */
+  final def reduce[AA >: A](implicit S: Semigroup[AA]): AA =
+    S.combineAllOption(iterator).get
+
+  /**
+    * Applies the supplied function to each element and returns a new NonEmptyLazyList from the concatenated results
+    */
+  final def flatMap[B](f: A => NonEmptyLazyList[B]): NonEmptyLazyList[B] =
+    create(toLazyList.flatMap(f.andThen(_.toLazyList)))
+
+  /**
+    * Returns the number of elements
+    */
+  final def length: Long = toLazyList.size
+
+  /**
+    * Zips this `NonEmptyLazyList` with another `NonEmptyLazyList` and applies a function for each pair of elements
+    */ // TODO manually implemented
+  final def zipWith[B, C](b: NonEmptyLazyList[B])(f: (A, B) => C): NonEmptyLazyList[C] =
+    create(toLazyList.zip(b.toLazyList).map { case (a, b) => f(a, b) })
+
+  /**
+    * Zips each element of this `NonEmptyLazyList` with its index
+    */
+  final def zipWithIndex: NonEmptyLazyList[(A, Int)] =
+    create(toLazyList.zipWithIndex)
+
+  /**
+    * Groups elements inside this `NonEmptyLazyList` according to the `Order`
+    * of the keys produced by the given mapping function.
+    */
+  final def groupBy[B](f: A => B): NonEmptyMap[B, NonEmptyLazyList[A]] =
+    toLazyList.groupBy(f).asInstanceOf[NonEmptyMap[B, NonEmptyLazyList[A]]]
+
+  final def iterator: Iterator[A] = toLazyList.iterator
+
+  final def reverseIterator: Iterator[A] = toLazyList.reverseIterator
+
+  /**
+    * Reverses this `NonEmptyLazyList`
+    */
+  final def reverse: NonEmptyLazyList[A] =
+    create(toLazyList.reverse)
+
+  /**
+    * Remove duplicates.
+    */
+  final def distinct: NonEmptyLazyList[A] =
+    create(toLazyList.distinct)
+}
+
+sealed abstract private[data] class NonEmptyLazyListInstances extends NonEmptyLazyListInstances1 {
```

Comment:
I am going to loop @LukaJCB since I am not sure I follow the implementation of instances in `NonEmptyChain`

Namely I think since this is just a new type, i.e. at runtime it's just exactly the same type as `LazyList`, we should take advantage of the existing instances of `LazyList`. 

We could just do 

``` scala
import cats.instances.lazyList._
sealed abstract private[data] class NonEmptyLazyListInstances  extends NonEmptyLazyListInstances1  {
   
   implicit val catsDataBimonadForNonEmptyLazyList: Bimonad[NonEmptyLazyList] = Bimonad[LazyList].asInstanceOf[Bimonad[NonEmptyLazyList]]

} 

sealed abstract private[data] class NonEmptyLazyListInstances1  extends NonEmptyLazyListInstances2  {

   implicit val catsDataSemigroupKForNonEmptyLazyList: SemigroupK[NonEmptyLazyList] = SemigroupK[LazyList].asInstanceOf[SemigroupK[NonEmptyLazyList]]
 
   implicit def catsDataPartialOrderForNonEmptyLazyList[A: PartialOrder]: PartialOrder[NonEmptyLazyList[A]] =
    PartialOrder[LazyList[A]].asInstanceOf[PartialOrder[NonEmptyLazyList]]
}

sealed abstract private[data] class NonEmptyLazyListInstances2 {
   implicit val catsDataNonEmptyTraverseForNonEmptyLazyList:  NonEmptyTraverse[LazyList] = new NonEmptyTraverse[LazyList] {
     val  reducibleInstance = Reducible[LazyList].asInstanceOf[Reducible[NonEmptyLazyList]]
     val  traverseInstance = Traverse[LazyList].asInstanceOf[Traverse[NonEmptyLazyList]]
   
     //delegate all reducible methods to reducibleInstance
     //delegate all traverse methods to traverseInstance

     def nonEmptyTraverse[G[_]: Apply, A, B](fa: NonEmptyLazyList[A])(f: A => G[B]): G[NonEmptyLazyList[B]] =
        Foldable[LazyList].reduceRightToOption[A, G[LazyList[B]]](fa.tail)(a => Apply[G].map(f(a))(Lazylist.one)) { (a, lglb) =>
          Apply[G].map2Eval(f(a), lglb)(_ +: _)
        }.map {
           case None        => Apply[G].map(f(fa.head))(NonEmptyLazyList.apply)
           case Some(gtail) => Apply[G].map2(f(fa.head), gtail)((h, t) => create(NonEmptyLazyList(h) ++ t))
        } .value
}

  implicit def catsDataEqForNonEmptyLazyList[A: Eq]: Eq[NonEmptyLazyList[A]] = Eq[LazyList[A]].asInstanceOf[PartialOrder[NonEmptyLazyList]]
}
```
Right?

Reply:
Note that I distribute the instances into different traits so that they don't conflict with each other

Reply:
also note that by "delegate all traverse methods to traverseInstance"  I meant all `Traverse` methods that are implemented or overriden in the `Traverse` instance for `LazyList`. They are mostly https://github.com/typelevel/cats/blob/master/core/src/main/scala-2.13+/cats/instances/lazyList.scala#L44-L171, except `tailRecM`

Reply:
Ok I didn't quite realize what this pattern was doing but this makes a lot of sense and it's _really_ cool.

Reply:
Looks like the only instances that aren't already included are `BiMonad` `SemigroupK` and `PartialOrder`. Although I'm not getting it to recognize `Traverse` `Foldable` even though they're defined.

Reply:
Commit 5cee9d3 basically just slaps your suggested code on.